File: iOS/Tests/AutomatedTestsUITests/Steps/TextInputSteps.swift
/*
* Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*/

import Foundation
import XCTest

// swiftlint:disable implicitly_unwrapped_optional
// swiftlint:disable force_unwrapping
class TextInputSteps: CucumberStepsDefinition {
    var application: XCUIApplication!

    func loadSteps() {
        // MARK: - Before
        before { scenarioDefinition in
            if scenarioDefinition?.tags.contains("textinput") ?? false {
                let url = "http://localhost:8080/textinput"
                self.application = TestUtils.launchBeagleApplication(url: url)
            }
        }

        // MARK: - Given
        Given(#"^the Beagle application did launch with the textInput on screen$"#) { _, _ -> Void in
            XCTAssertTrue(ScreenElements.TEXT_INPUT_SCREEN_TITLE.element.exists)
        }

        // MARK: - When
        
        // Scenarios 5, 6 and 8
        When(#"^I click in the textInput with the placeholder "([^\"]*)"$"#) { args, _ -> Void in
            let placeholder = args![0]

            if let textField = self.application.textFields[placeholder: placeholder] {
                self.application.scrollToElement(element: textField)
                textField.tap()
            }
        }

        // MARK: - Then
        
        // Scenario 1
        Then(#"^I must check if the textInput value "([^\"]*)" appears on the screen$"#) { args, _ -> Void in
            let text = args![0]
            XCTAssertTrue(self.application.textFields[text].exists)
        }

        // Scenario 2
        Then(#"^I must check if the textInput placeholder "([^\"]*)" appears on the screen$"#) { args, _ -> Void in
            let placeholder = args![0]
            let textField = self.application.textFields[placeholder: placeholder]

            XCTAssertTrue(textField?.exists ?? false)
        }

        // Scenario 3
        Then(#"^verify if the field with the placeholder "([^\"]*)" is disabled$"#) { args, _ -> Void in
            let placeholder = args![0]
            let textField = self.application.textFields[placeholder: placeholder]

            XCTAssertTrue(textField?.exists ?? false)
            XCTAssertFalse(textField?.isEnabled ?? true)
        }

        // Scenario 4
        Then(#"^verify if the field with the value "([^\"]*)" is read only$"#) { args, _ -> Void in
            let text = args![0]
            let textField = self.application.textFields[text]

            XCTAssertTrue(textField.exists)
            XCTAssertFalse(textField.isEnabled)
        }

        // Scenario 5
        Then(#"^verify if the textInput "([^\"]*)" is the first responder$"#) { args, _ -> Void in
            let placeholder = args![0]
            let textField = self.application.textFields[placeholder: placeholder]

            XCTAssertTrue(textField?.exists ?? false)
            XCTAssertEqual(self.application.keyboards.count, 1)
        }

        // Scenario 6
        Then(#"^validate textInput component of type number with text "([^\"]*)"$"#) { args, _ -> Void in
            let placeholder = args![0]
            let textField = self.application.textFields[placeholder: placeholder]

            XCTAssertTrue(textField?.exists ?? false)
            XCTAssertEqual(self.application.keyboards.count, 1)
        }

        // Scenario 7
        Then(#"^change to "([^\"]*)" then to "([^\"]*)" then to "([^\"]*)"$"#) { args, _ -> Void in
            let didOnFocus = args![0]
            let didOnChange = args![1]
            let didOnBlur = args![2]

            XCTAssertTrue(self.application.textFields[didOnFocus].exists)
            self.application.typeText("a")

            XCTAssertTrue(self.application.textFields[didOnChange].exists)
            self.application.typeText("\n")

            XCTAssertTrue(self.application.textFields[didOnBlur].exists)
        }

        // Scenario 8
        Then(#"^change to "([^\"]*)" then to "([^\"]*)" then to "([^\"]*)" in the correct order$"#) { args, _ -> Void in
            let didOnFocus = args![0]
            let didOnChange = args![1]
            let didOnBlur = args![2]

            XCTAssertTrue(self.application.textFields[didOnFocus].exists)
            self.application.typeText("a")

            XCTAssertTrue(self.application.textFields[didOnChange].exists)
            self.application.typeText("\n")

            XCTAssertTrue(self.application.textFields[didOnBlur].exists)

            XCTAssertFalse(self.application.textFields[didOnFocus].exists)
            XCTAssertFalse(self.application.textFields[didOnChange].exists)
        }

        // Scenario 9
        Then(#"^The hidden input fields "([^\"]*)" should not be visible$"#) { args, _ -> Void in
            let text = args![0]
            let staticText = self.application.staticTexts["There are two hidden input fields above"]

            self.application.scrollToElement(element: staticText)
            XCTAssertFalse(self.application.textFields[text].exists)
        }
    }
}

private extension XCUIElementQuery {
    subscript(placeholder value: String) -> XCUIElement? {
        first { $0.placeholderValue == value }
    }
}

File: android/automated-tests/src/androidTest/kotlin/br/com/zup/beagle/automatedTests/cucumber/steps/TextInputSteps.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.automatedTests.cucumber.steps

import androidx.test.rule.ActivityTestRule
import br.com.zup.beagle.automatedTests.activity.MainActivity
import br.com.zup.beagle.automatedTests.cucumber.robots.ScreenRobot
import br.com.zup.beagle.automatedTests.utils.ActivityFinisher
import br.com.zup.beagle.automatedTests.utils.TestUtils
import cucumber.api.java.After
import cucumber.api.java.Before
import cucumber.api.java.en.And
import cucumber.api.java.en.Given
import cucumber.api.java.en.Then
import cucumber.api.java.en.When
import org.junit.Rule

const val TEXT_INPUT_SCREEN_BFF_URL = "http://10.0.2.2:8080/textinput"

class TextInputScreen {
    @Rule
    var activityTestRule = ActivityTestRule(MainActivity::class.java)

    @Before("@textinput")
    fun setup() {
        TestUtils.startActivity(activityTestRule, TEXT_INPUT_SCREEN_BFF_URL)
    }

    @Given("^the Beagle application did launch with the textInput on screen$")
    fun checkBaseScreen() {
        ScreenRobot()
            .checkViewContainsText("Beagle Text Input", true)
    }

    @Then("^I must check if the textInput value (.*) appears on the screen$")
    fun checkTextInput(string: String) {
        ScreenRobot()
            .checkViewContainsText(string, true)
    }

    @Then("^I must check if the textInput placeholder (.*) appears on the screen$")
    fun checkTextInputPlaceholder(string: String) {
        ScreenRobot()
            .checkViewContainsHint(string, true)
    }

    @When("^the disabled textInput (.*) is visible$")
    fun checkIfTextInputIsDisabled(string: String) {
        ScreenRobot()
            .checkViewContainsHint(string)
    }

    @Then("^verify if it's (.*) disabled$")
    fun checkDisabledField(string: String) {
        ScreenRobot()
            .disabledFieldHint(string)
    }

    @When("^the value (.*) of the readOnly field is on the screen$")
    fun checkIfTextInputIsReadOnly(string: String) {
        ScreenRobot()
            .checkViewContainsText(string)
    }

    @Then("^verify if the field with the value (.*) is read only$")
    fun checkReadOnlyField(string: String) {
        ScreenRobot()
            .disabledFieldText(string)
    }

    @When("^I click in the textInput with the placeholder (.*)$")
    fun checkTextInputInSecondPlan(string: String) {
        ScreenRobot()
            .checkViewContainsHint(string)
            .clickOnInputWithHint(string)
    }

    @Then("^verify if the text (.*) is in the second plan$")
    fun checkKeyboardFocus(string: String) {
        ScreenRobot()
            .hintInSecondPlan(string)
    }

    @Then("^validate that a textInput (.*) of type number is set$")
    fun validateTextsInputNumberType(string: String) {
        ScreenRobot()
            .checkInputTypeNumber(string)
    }

    @And("^I click the textInput with the placeholder (.*)$")
    fun textInoutWithActionOfOnFocusAndOnFocus(string: String) {
        ScreenRobot()
            .scrollToWithHint(string)
            .clickOnInputWithHint(string)
    }

    @Then("^the textInput with the placeholder \"Ordered actions\" should have value (.*)$")
    fun checkOrderedActions(string: String) {
        ScreenRobot()
            .checkViewContainsText(string)
    }

    @Then("^the textInput with the placeholder \"Unordered actions\" will change its value to (.*)$")
    fun textInoutWithActionOfOnBlur(string: String) {
        ScreenRobot()
            .checkViewContainsText(string)
    }

    @And("^I type anything on textInput with the placeholder (.*)$")
    fun triggersOnChangeMethodAndCheckChanges(string: String) {
        ScreenRobot()
            .typeText(string, "a")
    }

    @When("^I click to textInput (.*) then change to (.*) and to (.*)$")
    fun textInoutWithActionOfOnFocusAndOnChange(string: String, string2: String, string3: String) {
        ScreenRobot()
            .scrollToWithHint(string)
            .checkViewContainsHint(string)
            .clickOnInputWithHint(string)
            .checkViewContainsText(string2)
            .typeText(string, "a")
            .checkViewContainsText(string3)
    }

    @Then("^the text (.*) should be appear in the correctly order$")
    fun textInoutWithActionOfOnBlurCorrectlyOrder(string: String) {
        ScreenRobot()
            .checkViewContainsHint("is textInput type number")
            .clickOnInputWithHint("is textInput type number")
            .checkViewContainsText(string)
            .checkViewDoesNotContainsText("DidOnFocus")
            .checkViewDoesNotContainsText("DidOnFocusDidOnChange")
    }

    @Then("^The hidden input fields (.*) should not be visible$")
    fun checkInputTextIsHidden(string: String) {
        ScreenRobot()
            .scrollTo("There are two hidden input fields above")
            .checkViewIsNotDisplayed(string)
    }

    @After("@textinput")
    fun tearDown() {
        ActivityFinisher.finishOpenActivities()
    }

}